File: libs/server/domain/src/domain/workspace.rs
use std::{collections::HashMap, sync::Arc};

use super::core::{Entity, HasMany, HasOne};
use super::diagram::{Diagram, WorkspaceDiagram};
use super::logical_entity::{LogicalEntity, WorkspaceLogicalEntities};
use super::logical_relationship::{LogicalRelationship, WorkspaceLogicalRelationships};
use super::member::{Member, WorkspaceMembers};

#[derive(Debug, Clone)]
pub struct WorkspaceDescription {
    pub title: String,
    pub description: Option<String>,
    pub status: String,
    pub metadata: HashMap<String, String>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Clone)]
pub struct Workspace {
    identity: String,
    description: WorkspaceDescription,
    members: Arc<dyn WorkspaceMembers>,
    diagram: Arc<dyn WorkspaceDiagram>,
    logical_entities: Arc<dyn WorkspaceLogicalEntities>,
    logical_relationships: Arc<dyn WorkspaceLogicalRelationships>,
}

impl Workspace {
    pub fn new(
        identity: String,
        description: WorkspaceDescription,
        members: Arc<dyn WorkspaceMembers>,
        diagram: Arc<dyn WorkspaceDiagram>,
        logical_entities: Arc<dyn WorkspaceLogicalEntities>,
        logical_relationships: Arc<dyn WorkspaceLogicalRelationships>,
    ) -> Self {
        Self {
            identity,
            description,
            members,
            diagram,
            logical_entities,
            logical_relationships,
        }
    }

    pub fn identity(&self) -> &str {
        &self.identity
    }

    pub fn description(&self) -> &WorkspaceDescription {
        &self.description
    }

    pub fn members(&self) -> &dyn HasMany<Member> {
        self.members.as_ref()
    }

    pub fn members_wide(&self) -> &dyn WorkspaceMembers {
        self.members.as_ref()
    }

    pub fn diagram(&self) -> &dyn HasOne<Diagram> {
        self.diagram.as_ref()
    }

    pub fn logical_entities(&self) -> &dyn HasMany<LogicalEntity> {
        self.logical_entities.as_ref()
    }

    pub fn logical_entities_wide(&self) -> &dyn WorkspaceLogicalEntities {
        self.logical_entities.as_ref()
    }

    pub fn logical_relationships(&self) -> &dyn HasMany<LogicalRelationship> {
        self.logical_relationships.as_ref()
    }

    pub fn logical_relationships_wide(&self) -> &dyn WorkspaceLogicalRelationships {
        self.logical_relationships.as_ref()
    }
}

impl Entity for Workspace {
    type Identity = str;
    type Description = WorkspaceDescription;

    fn identity(&self) -> &Self::Identity {
        &self.identity
    }

    fn description(&self) -> &Self::Description {
        &self.description
    }
}

File: apps/server-nest/src/domain/workspace/workspace.ts
import { Entity, HasMany } from '../core';
import {
  Diagram,
  DiagramDescription,
  DraftEdge,
  DraftNode,
  WorkspaceDiagrams,
} from '../diagram';
import {
  LogicalEntity,
  LogicalEntityDescription,
  WorkspaceLogicalEntities,
} from '../logical-entity';
import {
  LogicalRelationship,
  LogicalRelationshipDescription,
  WorkspaceLogicalRelationships,
} from '../logical-relationship';
import { Member, MemberDescription, WorkspaceMembers } from '../member';

export interface WorkspaceDescription {
  title: string;
  description: string | null;
  status: string;
  metadata: Record<string, string>;
  createdAt: string;
  updatedAt: string;
}

export class Workspace implements Entity<string, WorkspaceDescription> {
  constructor(
    private readonly id: string,
    private readonly desc: WorkspaceDescription,
    private readonly workspaceMembers: WorkspaceMembers,
    private readonly workspaceDiagrams: WorkspaceDiagrams,
    private readonly workspaceLogicalEntities: WorkspaceLogicalEntities,
    private readonly workspaceLogicalRelationships: WorkspaceLogicalRelationships,
  ) {}

  identity(): string {
    return this.id;
  }

  description(): WorkspaceDescription {
    return this.desc;
  }

  members(): HasMany<Member> {
    return this.workspaceMembers;
  }

  addMember(desc: MemberDescription): Promise<Member> {
    return this.workspaceMembers.addMember(desc);
  }

  removeMember(userId: string): Promise<void> {
    return this.workspaceMembers.removeMember(userId);
  }

  diagrams(): HasMany<Diagram> {
    return this.workspaceDiagrams;
  }

  addDiagram(desc: DiagramDescription): Promise<Diagram> {
    return this.workspaceDiagrams.add(desc);
  }

  updateDiagram(
    diagramId: string,
    desc: DiagramDescription,
  ): Promise<Diagram> {
    return this.workspaceDiagrams.update(diagramId, desc);
  }

  deleteDiagram(diagramId: string): Promise<void> {
    return this.workspaceDiagrams.delete(diagramId);
  }

  listDiagrams(
    page: number,
    pageSize: number,
  ): Promise<[Diagram[], number]> {
    return this.workspaceDiagrams.list(page, pageSize);
  }

  saveDiagram(
    diagramId: string,
    draftNodes: DraftNode[],
    draftEdges: DraftEdge[],
  ): Promise<void> {
    return this.workspaceDiagrams.saveDiagram(diagramId, draftNodes, draftEdges);
  }

  publishDiagram(diagramId: string): Promise<void> {
    return this.workspaceDiagrams.publishDiagram(diagramId);
  }

  logicalEntities(): HasMany<LogicalEntity> {
    return this.workspaceLogicalEntities;
  }

  addLogicalEntity(desc: LogicalEntityDescription): Promise<LogicalEntity> {
    return this.workspaceLogicalEntities.add(desc);
  }

  updateLogicalEntity(
    entityId: string,
    desc: LogicalEntityDescription,
  ): Promise<LogicalEntity> {
    return this.workspaceLogicalEntities.update(entityId, desc);
  }

  deleteLogicalEntity(entityId: string): Promise<void> {
    return this.workspaceLogicalEntities.delete(entityId);
  }

  listLogicalEntities(
    page: number,
    pageSize: number,
  ): Promise<[LogicalEntity[], number]> {
    return this.workspaceLogicalEntities.list(page, pageSize);
  }

  logicalRelationships(): HasMany<LogicalRelationship> {
    return this.workspaceLogicalRelationships;
  }

  addLogicalRelationship(
    desc: LogicalRelationshipDescription,
  ): Promise<LogicalRelationship> {
    return this.workspaceLogicalRelationships.add(desc);
  }

  updateLogicalRelationship(
    relationshipId: string,
    desc: LogicalRelationshipDescription,
  ): Promise<LogicalRelationship> {
    return this.workspaceLogicalRelationships.update(relationshipId, desc);
  }

  deleteLogicalRelationship(relationshipId: string): Promise<void> {
    return this.workspaceLogicalRelationships.delete(relationshipId);
  }

  listLogicalRelationships(
    page: number,
    pageSize: number,
  ): Promise<[LogicalRelationship[], number]> {
    return this.workspaceLogicalRelationships.list(page, pageSize);
  }
}
